qiskit/transpiler/passmanager.py

```diff
@@ -221,17 +224,21 @@ def callback_func(**kwargs):
                         property_set = kwargs['property_set']
                         count = kwargs['count']
                         ...
+            return_dag: If True return DAGCircuit as output instead of converting
+                        back to a QuantumCircuit.
```

Comment:
Having an explicit option that needs threading through many places feels a bit wonky to me, and may cause problems for Naoki's work on genericising the `PassManager` logic so it can be reused for pulse stuff.

Are there any major issues with us inferring the output type based on the input type?  With the additions you've made to `dag_to_circuit` above (which I agree are a good idea - `PassManager` shouldn't be setting magic properties that nowhere else knows about), if a user wants to do the `DAGCircuit -> PassManager -> QuantumCircuit` path, they'd easily be able to manually do `dag_to_circuit(pm.run(dag))` to make it explicit.

It might go easier if we move the `circuit_to_dag` conversion logic out of `RunningPassManager` and into `PassManager`, and then `RunningPassManager` can _solely_ worry about `DAGCircuit`s.  I'd even consider moving the logic that copies property-set stuff over from `RunningPassManager` into `PassManager` as well.  `RunningPassManager` is necessary to have an object that tracks the state of iteration through the passes and the flow control in a way that lets the outer `PassManager` restart / be thread-safe re-entrant / be mutated / etc (`PassManager` is to `RunningPassManager` approximately as `list` is to `list_iterator`), but the logic of the `PassManager.run` interface could easily be responsible for the initial/final conversion logic.  Even if we stuck with the kwarg to control the return value, in this form we wouldn't need to thread it through 5 API calls.

Reply:
I second the idea of inferring the output type and then explicitly converting to QC if desired. Here's some motivation:
The question of whether to infer (or otherwise choose) the return type or to select it via an argument arises in many contexts. I agree that if an object is passed through a complicated flow, then threading the choice through calls can be messy. The main argument I've come up with (I never studied this, I don't know how to search for the issue) for choosing the return type via the interface is that it's convenient if the interface is at a high level, more or less an application.  That's not the case here. Although QC and DC are not currently passed through many transforming functions, in comparison to say expressions in a CAS, it's reasonable to allow for this complexity to increase.